File: intersectionOfTwoLL.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

/*
 * Approach:
 * In order to find the intersection node, we first find the lengths
 * of the 2 LLs and then move one of the pointers by the difference of
 * those lengths, then we move both pointers at 1x speed until the common node
 * is found.
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
import java.util.*;
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null && headB == null) {
            return null;
        }

        int lenA = 0, lenB = 0;

        // find lengths of the 2 lists
        ListNode ptr = headA;
        while(ptr != null) {
            lenA++;
            ptr = ptr.next;
        }
        ptr = headB;
        while(ptr != null) {
            lenB++;
            ptr = ptr.next;
        }

        // then move head of the larger list forward by diff
        if(lenA > lenB) {
            for(int i=0; i < lenA-lenB && headA != null; i++) {
                headA = headA.next;
            }
        } else if(lenA < lenB) {
            for(int i=0; i < lenB-lenA && headB != null; i++) {
                headB = headB.next;
            }
        }

        // then find common node
        while(headA != null && headB != null && headA != headB) {
            headA = headA.next;
            headB = headB.next;
        }

        return headA;
    }
}

File: Intersection_of_Two_LL.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int LenA = 0, LenB=0;
        ListNode *curr = headA;
        while(curr!=NULL){
            curr = curr->next;
            LenA++;
        }
        curr = headB;
        while(curr!=NULL){
            curr = curr->next;
            LenB++;
        }
        while(LenA>LenB){
            headA = headA->next;
            LenA--;
        }
        while(LenB>LenA){
            headB = headB->next;
            LenB--;
        }
        while(headA!=headB){
          headA = headA->next;
          headB = headB->next;  
        }
        return headA;
    }
};


/*
class Solution {
    int Length_Calculator(ListNode* node){
        int length=0;
        while(node!=NULL){
            node = node->next;
            length++;
        }
        return length;
    }
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int l1=Length_Calculator(headA);
        int l2=Length_Calculator(headB);
        int d;
        ListNode*ptr1;
        ListNode*ptr2;

        if(l1>l2){
            d = l1-l2;
            ptr1 = headA;
            ptr2 = headB;
        }
        else{
            d= l2-l1;
            ptr1=headB;
            ptr2=headA;
        }
        while(d--){
            ptr1 = ptr1->next;
            if(ptr1==NULL)return NULL;
        }
        while(ptr1!=NULL && ptr2!=NULL){
            if(ptr1==ptr2) return ptr1;
            ptr1=ptr1->next;
            ptr2=ptr2->next;
        }
return NULL;        
    }
};


*/
